include/ideep/operators/matmul.hpp

```diff
@@ -818,23 +819,42 @@ struct matmul_forward : public dnnl::matmul,
     // introduces *an extra reorder* afterwards. Here we keep the weight format
     // untouched thanks to optimizations for both plain and transposed formats
     // in DNNL.
-    IDEEP_ENFORCE(weights.get_data_type() == data_type::f32 ||
-                  weights.get_data_type() == data_type::bf16 ||
-                  weights.get_data_type() == data_type::f16,
-                  "Incorrect data type in weights");
-    dst_data_type = src.get_data_type() == data_type::bf16 ?
-                    data_type::bf16 :
-                    ((src.get_data_type() == data_type::f16) ?
-                      data_type::f16 : data_type::f32);
-    src_desc = src.get_desc().to_type(dst_data_type);
+    IDEEP_ENFORCE(
+        weights.get_data_type() == data_type::f32 ||
+            weights.get_data_type() == data_type::bf16 ||
+            weights.get_data_type() == data_type::f16,
+        "Incorrect data type in weights");
 
 #ifdef __aarch64__
+
+    dst_data_type = src.get_data_type() == data_type::bf16
+        ? ((dst.get_data_type() == data_type::f32) ? data_type::f32
+                                                   : data_type::bf16)
+        : ((src.get_data_type() == data_type::f16) ? data_type::f16
+                                                   : data_type::f32);
```

Comment:
The concern is that how this would impact the original callers since it is a BC breaking changes. @yanbing-j can you double check the existing callers?

Reply:
In the current implementation, the existing callers are with same dtypes (src and dst), for example, https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp#L1531-L1543 by using reproducer 
```
mat1 = torch.randn(20, 30).to(torch.bfloat16)
mat2 = torch.randn(30, 30).to(torch.bfloat16)
torch.matmul(mat1, mat2)
```

Reply:
For a broader context, this will allow to pick up the right kernel from oneDNN for matmul where `src: bf16`, `wei:bf16`and `dst:f32`. 

https://github.com/pytorch/pytorch/blob/1185975c6e7f23f0cb318e47166fceebd2995c98/aten/src/ATen/native/CPUBlas.cpp#L400 
For the current Operator that will use this will be: `scaled_dot_product_attention`

Reply:
@aditew01 Sorry, I cannot find the calling of `mkldnn_bf16f32_gemm` in the main branch https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L405. Could you try to update your code base?

Reply:
> @aditew01 Sorry, I cannot find the calling of `mkldnn_bf16f32_gemm` in the main branch https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L405. Could you try to update your code base?

Hi @yanbing-j , the change which I referred to is in a PR (draft) which I've raised in PyTorch. It's draft and waits on specific onednn and ideep changes. 

 https://github.com/pytorch/pytorch/pull/140159